# 📌 Extracción

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
datos = pd.read_json(r'C:\Users\crist\Documents\Proyectos\TelecomX_LATAM\data\raw\TelecomX_Data.json')
datos.head()

,customerID,Churn,customer,phone,internet,account
0,0002-ORFBO,No,"{'gender': 'Female', 'SeniorCitizen': 0, 'Part...","{'PhoneService': 'Yes', 'MultipleLines': 'No'}","{'InternetService': 'DSL', 'OnlineSecurity': '...","{'Contract': 'One year', 'PaperlessBilling': '..."
1,0003-MKNFE,No,"{'gender': 'Male', 'SeniorCitizen': 0, 'Partne...","{'PhoneService': 'Yes', 'MultipleLines': 'Yes'}","{'InternetService': 'DSL', 'OnlineSecurity': '...","{'Contract': 'Month-to-month', 'PaperlessBilli..."
2,0004-TLHLJ,Yes,"{'gender': 'Male', 'SeniorCitizen': 0, 'Partne...","{'PhoneService': 'Yes', 'MultipleLines': 'No'}","{'InternetService': 'Fiber optic', 'OnlineSecu...","{'Contract': 'Month-to-month', 'PaperlessBilli..."
3,0011-IGKFF,Yes,"{'gender': 'Male', 'SeniorCitizen': 1, 'Partne...","{'PhoneService': 'Yes', 'MultipleLines': 'No'}","{'InternetService': 'Fiber optic', 'OnlineSecu...","{'Contract': 'Month-to-month', 'PaperlessBilli..."
4,0013-EXCHZ,Yes,"{'gender': 'Female', 'SeniorCitizen': 1, 'Part...","{'PhoneService': 'Yes', 'MultipleLines': 'No'}","{'InternetService': 'Fiber optic', 'OnlineSecu...","{'Contract': 'Month-to-month', 'PaperlessBilli..."


In [3]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


# 🔧 Transformación

In [4]:
df = pd.json_normalize(datos.to_dict(orient='records'))
df

,customerID,Churn,customer.gender,customer.SeniorCitizen,customer.Partner,customer.Dependents,customer.tenure,phone.PhoneService,phone.MultipleLines,internet.InternetService,...,internet.OnlineBackup,internet.DeviceProtection,internet.TechSupport,internet.StreamingTV,internet.StreamingMovies,account.Contract,account.PaperlessBilling,account.PaymentMethod,account.Charges.Monthly,account.Charges.Total
0,0002-ORFBO,No,Female,0,Yes,Yes,9,Yes,No,DSL,...,Yes,No,Yes,Yes,No,One year,Yes,Mailed check,65.60,593.3
1,0003-MKNFE,No,Male,0,No,No,9,Yes,Yes,DSL,...,No,No,No,No,Yes,Month-to-month,No,Mailed check,59.90,542.4
2,0004-TLHLJ,Yes,Male,0,No,No,4,Yes,No,Fiber optic,...,No,Yes,No,No,No,Month-to-month,Yes,Electronic check,73.90,280.85
3,0011-IGKFF,Yes,Male,1,Yes,No,13,Yes,No,Fiber optic,...,Yes,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,98.00,1237.85
4,0013-EXCHZ,Yes,Female,1,Yes,No,3,Yes,No,Fiber optic,...,No,No,Yes,Yes,No,Month-to-month,Yes,Mailed check,83.90,267.4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7262,9987-LUTYD,No,Female,0,No,No,13,Yes,No,DSL,...,No,No,Yes,No,No,One year,No,Mailed check,55.15,742.9
7263,9992-RRAMN,Yes,Male,0,Yes,No,22,Yes,Yes,Fiber optic,...,No,No,No,No,Yes,Month-to-month,Yes,Electronic check,85.10,1873.7
7264,9992-UJOEL,No,Male,0,No,No,2,Yes,No,DSL,...,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,50.30,92.75
7265,9993-LHIEB,No,Male,0,Yes,Yes,67,Yes,No,DSL,...,No,Yes,Yes,No,Yes,Two year,No,Mailed check,67.85,4627.65


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

In [6]:
nuevos_nombres = {
    'customerID': 'customer_id',
    'Churn': 'churn',
    'customer.gender': 'gender',
    'customer.SeniorCitizen': 'senior_citizen',
    'customer.Partner': 'partner',
    'customer.Dependents': 'dependents',
    'customer.tenure': 'tenure',
    'phone.PhoneService': 'phone_service',
    'phone.MultipleLines': 'multiple_lines',
    'internet.InternetService': 'internet_service',
    'internet.OnlineSecurity': 'online_security',
    'internet.OnlineBackup': 'online_backup',
    'internet.DeviceProtection': 'device_protection',
    'internet.TechSupport': 'tech_support',
    'internet.StreamingTV': 'streaming_tv',
    'internet.StreamingMovies': 'streaming_movies',
    'account.Contract': 'contract',
    'account.PaperlessBilling': 'paperless_billing',
    'account.PaymentMethod': 'payment_method',
    'account.Charges.Monthly' : 'monthly_charges',
    'account.Charges.Total' : 'total_charges',
}

df = df.rename(columns=nuevos_nombres, errors='raise')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customer_id        7267 non-null   object 
 1   churn              7267 non-null   object 
 2   gender             7267 non-null   object 
 3   senior_citizen     7267 non-null   int64  
 4   partner            7267 non-null   object 
 5   dependents         7267 non-null   object 
 6   tenure             7267 non-null   int64  
 7   phone_service      7267 non-null   object 
 8   multiple_lines     7267 non-null   object 
 9   internet_service   7267 non-null   object 
 10  online_security    7267 non-null   object 
 11  online_backup      7267 non-null   object 
 12  device_protection  7267 non-null   object 
 13  tech_support       7267 non-null   object 
 14  streaming_tv       7267 non-null   object 
 15  streaming_movies   7267 non-null   object 
 16  contract           7267 

In [7]:
df.tail()

,customer_id,churn,gender,senior_citizen,partner,dependents,tenure,phone_service,multiple_lines,internet_service,...,online_backup,device_protection,tech_support,streaming_tv,streaming_movies,contract,paperless_billing,payment_method,monthly_charges,total_charges
7262,9987-LUTYD,No,Female,0,No,No,13,Yes,No,DSL,...,No,No,Yes,No,No,One year,No,Mailed check,55.15,742.9
7263,9992-RRAMN,Yes,Male,0,Yes,No,22,Yes,Yes,Fiber optic,...,No,No,No,No,Yes,Month-to-month,Yes,Electronic check,85.10,1873.7
7264,9992-UJOEL,No,Male,0,No,No,2,Yes,No,DSL,...,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,50.30,92.75
7265,9993-LHIEB,No,Male,0,Yes,Yes,67,Yes,No,DSL,...,No,Yes,Yes,No,Yes,Two year,No,Mailed check,67.85,4627.65
7266,9995-HOTOH,No,Male,0,Yes,Yes,63,No,No phone service,DSL,...,Yes,Yes,No,Yes,Yes,Two year,No,Electronic check,59.00,3707.6


In [8]:
df[df['churn'] == 'No']

,customer_id,churn,gender,senior_citizen,partner,dependents,tenure,phone_service,multiple_lines,internet_service,...,online_backup,device_protection,tech_support,streaming_tv,streaming_movies,contract,paperless_billing,payment_method,monthly_charges,total_charges
0,0002-ORFBO,No,Female,0,Yes,Yes,9,Yes,No,DSL,...,Yes,No,Yes,Yes,No,One year,Yes,Mailed check,65.60,593.3
1,0003-MKNFE,No,Male,0,No,No,9,Yes,Yes,DSL,...,No,No,No,No,Yes,Month-to-month,No,Mailed check,59.90,542.4
5,0013-MHZWF,No,Female,0,No,Yes,9,Yes,No,DSL,...,No,No,Yes,Yes,Yes,Month-to-month,Yes,Credit card (automatic),69.40,571.45
6,0013-SMEOE,No,Female,1,Yes,No,71,Yes,No,Fiber optic,...,Yes,Yes,Yes,Yes,Yes,Two year,Yes,Bank transfer (automatic),109.70,7904.25
7,0014-BMAQU,No,Male,0,Yes,No,63,Yes,Yes,Fiber optic,...,No,No,Yes,No,No,Two year,Yes,Credit card (automatic),84.65,5377.8
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7259,9979-RGMZT,No,Female,0,No,No,7,Yes,No,Fiber optic,...,Yes,No,No,Yes,Yes,One year,Yes,Mailed check,94.05,633.45
7262,9987-LUTYD,No,Female,0,No,No,13,Yes,No,DSL,...,No,No,Yes,No,No,One year,No,Mailed check,55.15,742.9
7264,9992-UJOEL,No,Male,0,No,No,2,Yes,No,DSL,...,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,50.30,92.75
7265,9993-LHIEB,No,Male,0,Yes,Yes,67,Yes,No,DSL,...,No,Yes,Yes,No,Yes,Two year,No,Mailed check,67.85,4627.65


In [9]:
df[(df['churn'] == '') & (df['tenure'] == 0)]

,customer_id,churn,gender,senior_citizen,partner,dependents,tenure,phone_service,multiple_lines,internet_service,...,online_backup,device_protection,tech_support,streaming_tv,streaming_movies,contract,paperless_billing,payment_method,monthly_charges,total_charges


In [10]:
df[df['tenure'] == 0]

,customer_id,churn,gender,senior_citizen,partner,dependents,tenure,phone_service,multiple_lines,internet_service,...,online_backup,device_protection,tech_support,streaming_tv,streaming_movies,contract,paperless_billing,payment_method,monthly_charges,total_charges
975,1371-DWPAZ,No,Female,0,Yes,Yes,0,No,No phone service,DSL,...,Yes,Yes,Yes,Yes,No,Two year,No,Credit card (automatic),56.05,
1775,2520-SGTTA,No,Female,0,Yes,Yes,0,Yes,No,No,...,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.00,
1955,2775-SEFEE,No,Male,0,No,Yes,0,Yes,Yes,DSL,...,Yes,No,Yes,No,No,Two year,Yes,Bank transfer (automatic),61.90,
2075,2923-ARZLG,No,Male,0,Yes,Yes,0,Yes,No,No,...,No internet service,No internet service,No internet service,No internet service,No internet service,One year,Yes,Mailed check,19.70,
2232,3115-CZMZD,No,Male,0,No,Yes,0,Yes,No,No,...,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.25,
2308,3213-VVOLG,No,Male,0,Yes,Yes,0,Yes,Yes,No,...,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.35,
2930,4075-WKNIU,No,Female,0,Yes,Yes,0,Yes,Yes,DSL,...,Yes,Yes,Yes,Yes,No,Two year,No,Mailed check,73.35,
3134,4367-NUYAO,No,Male,0,Yes,Yes,0,Yes,Yes,No,...,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.75,
3203,4472-LVYGI,No,Female,0,Yes,Yes,0,No,No phone service,DSL,...,No,Yes,Yes,Yes,No,Two year,Yes,Bank transfer (automatic),52.55,
4169,5709-LVOEQ,No,Female,0,Yes,Yes,0,Yes,No,DSL,...,Yes,Yes,No,Yes,Yes,Two year,No,Mailed check,80.85,


In [11]:
df['senior_citizen'].unique()

array([0, 1])

In [12]:
df['payment_method'].unique()

array(['Mailed check', 'Electronic check', 'Credit card (automatic)',
       'Bank transfer (automatic)'], dtype=object)

In [13]:
df[df['tenure'] < 12]['churn'].value_counts()

churn
No     1070
Yes     999
         77
Name: count, dtype: int64

In [14]:
columnas_de_texto = df.select_dtypes(include='object').columns

for col in columnas_de_texto:
    n = (df[col].str.strip() == '').sum()
    print(f'{col:20} {n}')

customer_id          0
churn                224
gender               0
partner              0


dependents           0
phone_service        0
multiple_lines       0
internet_service     0
online_security      0
online_backup        0
device_protection    0
tech_support         0
streaming_tv         0
streaming_movies     0
contract             0
paperless_billing    0
payment_method       0
total_charges        11


In [15]:
df['customer_id'].duplicated().sum()

np.int64(0)

In [16]:
df[df['churn'] == ''].value_counts().sum()

np.int64(224)

In [17]:
sin_target = df[df['churn'] == '']
con_target = df[df['churn'] != '']

print(f"Sin target: {len(sin_target)} | Con target: {len(con_target)}")

for col in ['contract', 'payment_method', 'internet_service']:
    comparacion = pd.DataFrame({
        'sin_target_%': sin_target[col].value_counts(normalize=True) * 100,
        'con_target_%': con_target[col].value_counts(normalize=True) * 100
    }).round(2)
    comparacion['diferencia'] = (comparacion['sin_target_%'] - comparacion['con_target_%']).round(2)
    print(f"\n--- {col} ---")
    print(comparacion)

print("\n--- tenure ---")
print(pd.DataFrame({
    'sin_target': sin_target['tenure'].describe(),
    'con_target': con_target['tenure'].describe()
}).round(2))

Sin target: 224 | Con target: 7043

--- contract ---
                sin_target_%  con_target_%  diferencia
contract                                              
Month-to-month         58.04         55.02        3.02
Two year               21.43         24.07       -2.64
One year               20.54         20.91       -0.37

--- payment_method ---
                           sin_target_%  con_target_%  diferencia
payment_method                                                   
Bank transfer (automatic)         20.09         21.92       -1.83
Credit card (automatic)           20.54         21.61       -1.07
Electronic check                  35.71         33.58        2.13
Mailed check                      23.66         22.89        0.77

--- internet_service ---
                  sin_target_%  con_target_%  diferencia
internet_service                                        
Fiber optic              45.54         43.96        1.58
DSL                      29.91         34.37       -4.4

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customer_id        7267 non-null   object 
 1   churn              7267 non-null   object 
 2   gender             7267 non-null   object 
 3   senior_citizen     7267 non-null   int64  
 4   partner            7267 non-null   object 
 5   dependents         7267 non-null   object 
 6   tenure             7267 non-null   int64  
 7   phone_service      7267 non-null   object 
 8   multiple_lines     7267 non-null   object 
 9   internet_service   7267 non-null   object 
 10  online_security    7267 non-null   object 
 11  online_backup      7267 non-null   object 
 12  device_protection  7267 non-null   object 
 13  tech_support       7267 non-null   object 
 14  streaming_tv       7267 non-null   object 
 15  streaming_movies   7267 non-null   object 
 16  contract           7267 

In [ ]:
print('-- Valores unicos de los servicios --')

print(f'\nValores unicos de multiple_lines:    {df["multiple_lines"].unique()}')
print(f'Valores unicos de online_security:   {df["online_security"].unique()}')
print(f'Valores unicos de online_backup:     {df["online_backup"].unique()}')
print(f'Valores unicos de device_protection: {df["device_protection"].unique()}')
print(f'Valores unicos de tech_support:      {df["tech_support"].unique()}')
print(f'Valores unicos de streaming_tv:      {df["streaming_tv"].unique()}')
print(f'Valores unicos de streaming_movies:  {df["streaming_movies"].unique()}')

-- Valores unicos de los servicios --
Valores unicos de multiple_lines:    ['No' 'Yes' 'No phone service']
Valores unicos de online_security:   ['No' 'Yes' 'No internet service']
Valores unicos de online_backup:     ['Yes' 'No' 'No internet service']
Valores unicos de device_protection: ['No' 'Yes' 'No internet service']
Valores unicos de tech_support:      ['Yes' 'No' 'No internet service']
Valores unicos de streaming_tv:      ['Yes' 'No' 'No internet service']
Valores unicos de streaming_movies:  ['No' 'Yes' 'No internet service']


In [20]:
df_clean = df.copy()
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customer_id        7267 non-null   object 
 1   churn              7267 non-null   object 
 2   gender             7267 non-null   object 
 3   senior_citizen     7267 non-null   int64  
 4   partner            7267 non-null   object 
 5   dependents         7267 non-null   object 
 6   tenure             7267 non-null   int64  
 7   phone_service      7267 non-null   object 
 8   multiple_lines     7267 non-null   object 
 9   internet_service   7267 non-null   object 
 10  online_security    7267 non-null   object 
 11  online_backup      7267 non-null   object 
 12  device_protection  7267 non-null   object 
 13  tech_support       7267 non-null   object 
 14  streaming_tv       7267 non-null   object 
 15  streaming_movies   7267 non-null   object 
 16  contract           7267 

In [21]:
df_clean = df_clean[df_clean['churn'] != '']

In [22]:
len(df_clean)

7043

In [23]:
df_clean['total_charges'] = df_clean['total_charges'].replace(' ', 0).astype(float)

In [33]:
df_clean['senior_citizen'] = df_clean['senior_citizen'].replace({0: 'No', 1: 'Yes'})
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7043 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customer_id        7043 non-null   object 
 1   churn              7043 non-null   object 
 2   gender             7043 non-null   object 
 3   senior_citizen     7043 non-null   object 
 4   partner            7043 non-null   object 
 5   dependents         7043 non-null   object 
 6   tenure             7043 non-null   int64  
 7   phone_service      7043 non-null   object 
 8   multiple_lines     7043 non-null   object 
 9   internet_service   7043 non-null   object 
 10  online_security    7043 non-null   object 
 11  online_backup      7043 non-null   object 
 12  device_protection  7043 non-null   object 
 13  tech_support       7043 non-null   object 
 14  streaming_tv       7043 non-null   object 
 15  streaming_movies   7043 non-null   object 
 16  contract           7043 non-n

In [36]:
df_clean['senior_citizen'].value_counts()

senior_citizen
No     5901
Yes    1142
Name: count, dtype: int64

In [37]:
servicios = ['online_security',
            'online_backup', 
            'device_protection',
            'tech_support',
            'streaming_tv', 
            'streaming_movies']

df_clean[servicios] = df_clean[servicios].replace('No internet service', 'No')
df_clean['service_count'] = (df_clean[servicios] == 'Yes').sum(axis=1)

In [39]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7043 entries, 0 to 7266
Data columns (total 22 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customer_id        7043 non-null   object 
 1   churn              7043 non-null   object 
 2   gender             7043 non-null   object 
 3   senior_citizen     7043 non-null   object 
 4   partner            7043 non-null   object 
 5   dependents         7043 non-null   object 
 6   tenure             7043 non-null   int64  
 7   phone_service      7043 non-null   object 
 8   multiple_lines     7043 non-null   object 
 9   internet_service   7043 non-null   object 
 10  online_security    7043 non-null   object 
 11  online_backup      7043 non-null   object 
 12  device_protection  7043 non-null   object 
 13  tech_support       7043 non-null   object 
 14  streaming_tv       7043 non-null   object 
 15  streaming_movies   7043 non-null   object 
 16  contract           7043 non-n

In [50]:
df_clean[df_clean['internet_service'] != 'No']

,customer_id,churn,gender,senior_citizen,partner,dependents,tenure,phone_service,multiple_lines,internet_service,...,device_protection,tech_support,streaming_tv,streaming_movies,contract,paperless_billing,payment_method,monthly_charges,total_charges,service_count
0,0002-ORFBO,No,Female,No,Yes,Yes,9,Yes,No,DSL,...,No,Yes,Yes,No,One year,Yes,Mailed check,65.60,593.30,3
1,0003-MKNFE,No,Male,No,No,No,9,Yes,Yes,DSL,...,No,No,No,Yes,Month-to-month,No,Mailed check,59.90,542.40,1
2,0004-TLHLJ,Yes,Male,No,No,No,4,Yes,No,Fiber optic,...,Yes,No,No,No,Month-to-month,Yes,Electronic check,73.90,280.85,1
3,0011-IGKFF,Yes,Male,Yes,Yes,No,13,Yes,No,Fiber optic,...,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,98.00,1237.85,4
4,0013-EXCHZ,Yes,Female,Yes,Yes,No,3,Yes,No,Fiber optic,...,No,Yes,Yes,No,Month-to-month,Yes,Mailed check,83.90,267.40,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7262,9987-LUTYD,No,Female,No,No,No,13,Yes,No,DSL,...,No,Yes,No,No,One year,No,Mailed check,55.15,742.90,2
7263,9992-RRAMN,Yes,Male,No,Yes,No,22,Yes,Yes,Fiber optic,...,No,No,No,Yes,Month-to-month,Yes,Electronic check,85.10,1873.70,1
7264,9992-UJOEL,No,Male,No,No,No,2,Yes,No,DSL,...,No,No,No,No,Month-to-month,Yes,Mailed check,50.30,92.75,1
7265,9993-LHIEB,No,Male,No,Yes,Yes,67,Yes,No,DSL,...,Yes,Yes,No,Yes,Two year,No,Mailed check,67.85,4627.65,4


In [51]:
df_clean.to_csv(r'C:\Users\crist\Documents\Proyectos\TelecomX_LATAM\data\processed\TelecomX_Data_clean.csv', index=False)

## Bitácora de decisiones — Fase 3

### Decisión 1 — Los 224 registros sin churn

**Qué encontré.** 224 registros (3,08% del total) tienen la columna `churn` vacía. No están como nulos, están como texto vacío `''`. Por eso `isna()` e `info()` no los detectan: para pandas, un texto vacío no es un nulo.

**Qué opciones tenía.** Dejarlos en todo el análisis. Dejarlos solo para describir la cartera y sacarlos de los cruces contra churn. O sacarlos del análisis.

**Qué elegí.** Sacarlos. El análisis queda con 7.043 registros, y ese es el número base de todo el proyecto.

**Por qué.** El churn es la variable que quiero explicar. Si le invento un valor a un registro que no lo tiene, estoy inventando el resultado. Un target no se imputa.

Descarté la opción intermedia porque me dejaba dos números distintos dando vueltas por el mismo informe, y es fácil confundirse al leerlo.

Antes de sacarlos revisé que no fueran un grupo raro. Comparé los 224 contra los 7.043 en contrato, método de pago, tipo de internet y antigüedad. La diferencia más grande fue de 4,46 puntos porcentuales (DSL), el resto por debajo de 3. La mediana de antigüedad da igual en los dos grupos: 29 meses. O sea, se reparten como el resto de la cartera y sacarlos no deforma la muestra.

Igual no desaparecen del proyecto: que falte el churn en el 3% de los registros es un problema de calidad de datos y va al informe.

**Qué asumo.** Que el churn falta por un fallo al registrar el dato, y no por algo que tenga que ver con el cliente. La comparación apoya esto pero no lo prueba: solo revisé cuatro variables. Si esos 224 fueran, por ejemplo, clientes cuyo registro se rompió justo cuando se estaban dando de baja, mi tasa de churn saldría más baja de lo real. Con los datos que tengo no lo puedo descartar.

---

### Decisión 2 — Los 11 registros sin `total_charges`

**Qué encontré.** 11 registros (0,15%) tienen `total_charges` vacío. Los 11 tienen `tenure = 0` y todos sí tienen `monthly_charges`. Por culpa de esos vacíos, la columna entera figuraba como texto. Ninguno de los 11 estaba entre los 224 que saqué en la decisión anterior.

**Qué opciones tenía.** Dejarlos como nulo. Ponerles 0. O borrar los registros.

**Qué elegí.** Ponerles 0 y convertir la columna a número.

**Por qué.** No estoy imputando, estoy deduciendo. Imputar es estimar un valor cuando no sabes cuál es. Acá sí lo sé: si el cliente lleva 0 meses, no ha acumulado nada. El valor sale del propio dato.

Poner nulo sería decir "no sé cuánto pagó", y sí lo sé. Además tiene un efecto práctico: la columna se quedaría como texto y esos 11 clientes desaparecerían sin avisar de cualquier promedio o gráfico de `total_charges` más adelante.

Borrarlos era lo más destructivo para un valor que puedo deducir.

**Nota al margen (no es un hallazgo).** De los 11, hay 10 con contrato de dos años y 1 con contrato de un año. Con 11 registros no se puede concluir nada, así que esto no va al informe. Lo dejo anotado nada más: son clientes que acaban de firmar el contrato más largo que ofrece la empresa.

**Qué asumo.** Que `tenure` está bien registrado, y que `tenure = 0` significa cliente nuevo al que todavía no le han facturado, no un error en esa columna.

**Qué cambia para después.** `total_charges` ahora tiene mínimo 0. Esos 11 van a aparecer en el extremo bajo de cualquier gráfico de esa variable.

---

### Decisión 3 — `senior_citizen` en 0/1

**Qué encontré.** `senior_citizen` viene como número entero (0 y 1), mientras que las demás columnas binarias del dataset usan `Yes`/`No`.

**Qué opciones tenía.** Dejarla como viene, o pasarla a `Yes`/`No`.

**Qué elegí.** Pasarla a texto: `1 → Yes`, `0 → No`. Antes revisé que la columna solo tuviera esos dos valores.

**Por qué.** Al ser número, aparece en `describe()` al lado de `tenure`, `monthly_charges` y `total_charges`, y su media se lee como si fuera un promedio. Pero no es un promedio, es una proporción. Eso confunde a quien lea el resumen.

Esto va un poco en contra de lo que decidí en la Fase 1, cuando dejé los nombres de columna en inglés para no alejarme de la fuente. Sigo con ese criterio, pero con un límite: si la fuente clasifica mal la variable, ahí sí la corrijo. `senior_citizen` es categórica, no numérica.

**Qué asumo.** Que 0 es No y 1 es Yes, que es lo habitual. No hay documentación de la fuente que lo diga.

---

### Decisión 4 — Variable nueva: `service_count`

**Para qué la creo.** Para responder esto: ¿los clientes que contratan más servicios adicionales se quedan más tiempo? Ninguna de las seis columnas de servicios responde eso por separado. Cada una me dice si tiene ese servicio puntual, pero no qué tan enganchado está el cliente con la empresa.

**Cómo la construí.** Cuento cuántos `Yes` tiene cada cliente en seis columnas: `online_security`, `online_backup`, `device_protection`, `tech_support`, `streaming_tv` y `streaming_movies`. Da un número de 0 a 6.

**Por qué esas seis y no otras.** Son las que el cliente puede elegir contratar o no. `phone_service` e `internet_service` quedan fuera porque son el servicio base: casi todos tienen alguno, así que no sirven para diferenciar. `multiple_lines` también es un adicional, pero la dejé fuera para que las seis midan lo mismo: todas dependen de tener internet contratado, mientras que `multiple_lines` depende de tener teléfono.

**Qué cambié antes de contar.** En las seis columnas convertí `No internet service` en `No`. Si el cliente no tiene internet, no puede tener el adicional, así que en la práctica es lo mismo que no tenerlo. Este cambio pisa las columnas originales y no se puede deshacer, pero no pierdo información: quién no tiene internet se sabe mirando `internet_service`.

**Cómo verifiqué que quedó bien.** Los valores van de 0 a 6, ninguno se sale. Y el control importante: todos los clientes sin internet quedaron con `service_count = 0`.

**Un problema que ya detecté y que afecta a la Fase 5.** El grupo de `service_count = 0` tiene 2.219 clientes, pero no son todos iguales. 1.526 no tienen internet, o sea que nunca pudieron contratar adicionales. Los otros 693 sí tienen internet y decidieron no contratar ninguno. Es el mismo número con dos significados opuestos.

Por eso, cuando cruce `service_count` contra churn, voy a usar solo los clientes que tienen internet: 5.517 registros, con el grupo 0 reducido a 693. El número lo voy a dejar escrito en el gráfico para que se vea que ahí el denominador cambia.

Los 1.526 sin internet siguen en todos los demás cruces. Y su comportamiento lo miro aparte, cruzando `internet_service` contra churn.

## Hipótesis previas al análisis

*Escritas el 2026-08-04, antes de ejecutar el EDA. No se modifican después.*

**1. — Tipo de contrato.**
Los clientes con contrato mes a mes abandonan más que los de contrato largo.
Mecanismo: el contrato mes a mes permite cancelar en cualquier momento, mientras
que los contratos largos obligan a esperar hasta el vencimiento. La retención del
contrato largo sería una barrera de salida, no necesariamente mayor satisfacción.

**2. — Método de pago.**
Los clientes con métodos de pago manuales abandonan más que los de pago automático.
Mecanismo: cada cobro manual es una decisión activa del cliente y una oportunidad
de no pagar. El pago automático sigue corriendo por inercia.

**3. — Antigüedad.**
Los clientes con menos de 12 meses de permanencia abandonan más que los de mayor
antigüedad.
Mecanismo: insatisfacción temprana con el servicio o aparición de mejores ofertas
de la competencia, sumado a menor fricción de salida en los primeros meses.
Nota: el umbral de 12 meses es una elección propia para hacer la hipótesis
contrastable, no un corte derivado del negocio. En la Fase 5 se verifica si el
quiebre real está en ese punto o en otro.

## Lista de diagnóstico

*Fase 2 — Auditoría previa a cualquier corrección. Ningún dato ha sido modificado.*

**1. Vacíos en el target**

La columna `churn` presenta 224 registros con cadena vacía `''`, equivalente al 3,08%
de los 7.267 registros. Estos valores no son detectados por `isna()` ni por `info()`:
pandas solo considera nulos a `NaN`, `None` y `NaT`, mientras que `''` es una cadena
de texto válida de longitud cero. Fue necesaria una comprobación explícita por
comparación directa para encontrarlos.

**2. `total_charges` no numérica**

La columna llega con dtype `object` en lugar de `float64`. La causa son 11 registros
(0,15%) que quedan vacíos al eliminar espacios, lo que impide la conversión a
numérico: basta un valor no convertible para degradar el tipo de toda la columna.
Los 11 registros tienen `tenure = 0`, por lo que la ausencia es estructuralmente
coherente — sin meses de permanencia no hay cargo acumulado que registrar.
Contraste que lo confirma: `monthly_charges`, del mismo bloque y con el mismo
formato de origen, sí convirtió a `float64`.

**3. Inconsistencia de codificación**

`senior_citizen` viene como `int64` (0/1), mientras que el resto de variables
binarias del dataset usan texto (`Yes`/`No`).

**4. Duplicados — verificado, sin hallazgos**

`customer_id` no presenta ningún valor repetido. Al ser el identificador único por
cliente, su unicidad hace imposible la existencia de filas idénticas en las 21
columnas, por lo que ambos casos quedan descartados con una sola verificación.

**5. Relación no aritmética entre cargos** *(nota de contexto, no es un error)*

`monthly_charges × tenure` no reproduce el valor de `total_charges`. Las variables
sí están relacionadas —el acumulado depende del cargo y de la antigüedad— pero no
existe equivalencia aritmética exacta: `monthly_charges` es la tarifa vigente hoy,
mientras que `total_charges` es el acumulado histórico. Cambios de plan,
promociones o ajustes de tarifa rompen la identidad. Se documenta como
característica del dataset, no como inconsistencia a corregir.

**Verificación adicional:** las 18 columnas de texto fueron revisadas con
`.str.strip()`. Únicamente `churn` y `total_charges` presentan valores vacíos.

# 📊 Carga y análisis

#📄Informe final